File: core/api/typescript/src/module_launch_api/args_io.ts

// All module containers accept exactly one environment variable, which contains the serialized params that

import log = require("loglevel");
import { err, ok, Result } from "neverthrow";
import { ModuleContainerArgs } from "./module_container_args";

// dictate how the module container ought to behave
const SERIALIZED_ARGS_ENV_VAR = "SERIALIZED_ARGS";
const JAVASCRIPT_STRING_VAR_TYPE = "string";

const safeJsonParse = Result.fromThrowable(JSON.parse, (value: unknown): Error => {
    if (value instanceof Error) {
        return value;
    }
    return new Error("Received an unknown exception value that wasn't an error: " + value);
});

// Intended to be used when starting the container - gets the environment variables that the container should be started with
export function getEnvFromArgs(args: ModuleContainerArgs): Result<Map<string, string>, Error> {
    let argsStr: string;
    try {
        argsStr = JSON.stringify(args);
    } catch (jsonErr: any) {
        // Sadly, we have to do this because there's no great way to enforce the caught thing being an error
        // See: https://stackoverflow.com/questions/30469261/checking-for-typeof-error-in-js
        if (jsonErr && jsonErr.stack && jsonErr.message) {
            return err(jsonErr as Error);
        }
        return err(new Error("Stringify-ing ModuleContainerArgs object threw an exception, but " +
            "it's not an Error so we can't report any more information than this"));
    }
    const result = new Map<string, string>();
    result.set(SERIALIZED_ARGS_ENV_VAR, argsStr);
    return ok(result);
}

export function getArgsFromEnv(): Result<ModuleContainerArgs, Error> {
    const serializedParamsStr = process.env[SERIALIZED_ARGS_ENV_VAR];
    if (serializedParamsStr === undefined) {
		return err(new Error(`No serialized args variable '${SERIALIZED_ARGS_ENV_VAR}' defined`))
    }
    if (serializedParamsStr === "") {
		return err(new Error(`Found serialized args environment variable '${SERIALIZED_ARGS_ENV_VAR}', but the value was empty`))
    }
    const deserializationResult = safeJsonParse(serializedParamsStr);
    if (deserializationResult.isErr()) {
        // TODO Replace with a stacktrace.Propagate equivalent
        log.error(`An error occurred deserializing the args JSON '${serializedParamsStr}'`);
        return err(deserializationResult.error);
    }
    const args = deserializationResult.value as ModuleContainerArgs;

	// Generic validation based on field type
    for (let [key, value] of Object.entries(args)) {
        // Ensure no empty strings
        if (typeof value === JAVASCRIPT_STRING_VAR_TYPE && (value as string).trim() === "") {
            return err(new Error(`JSON field '${key}' is whitespace or empty string`));
        }
    }

    return ok(args);
}

File: core/api/golang/module_launch_api/args_io.go
/*
 * Copyright (c) 2021 - present Kurtosis Technologies Inc.
 * All Rights Reserved.
 */

package module_launch_api

import (
	"encoding/json"
	"github.com/kurtosis-tech/stacktrace"
	"os"
	"reflect"
	"strings"
)

const (
	// All module containers accept exactly one environment variable, which contains the serialized params that
	// dictate how the module container ought to behave
	serializedArgsEnvVar = "SERIALIZED_ARGS"

	jsonFieldTag          = "json"
)

// Intended to be used when starting the container - gets the environment variables that the container should be started with
func GetEnvFromArgs(args *ModuleContainerArgs) (map[string]string, error) {
	argsBytes, err := json.Marshal(args)
	if err != nil {
		return nil, stacktrace.Propagate(err, "An error occurred serializing module container args to JSON")
	}

	argsStr := string(argsBytes)
	return map[string]string{
		serializedArgsEnvVar: argsStr,
	}, nil
}

// Intended to be used in the container main.go function - gets args from the environment
func GetArgsFromEnv() (*ModuleContainerArgs, error) {
	serializedParamsStr, found := os.LookupEnv(serializedArgsEnvVar)
	if !found {
		return nil, stacktrace.NewError("No serialized args variable '%v' defined", serializedArgsEnvVar)
	}
	if serializedParamsStr == "" {
		return nil, stacktrace.NewError("Found serialized args environment variable '%v', but the value was empty", serializedArgsEnvVar)
	}
	paramsJsonBytes := []byte(serializedParamsStr)
	var args ModuleContainerArgs
	if err := json.Unmarshal(paramsJsonBytes, &args); err != nil {
		return nil, stacktrace.Propagate(err,"An error occurred deserializing the args JSON '%v'", serializedParamsStr)
	}

	// Generic validation based on field type
	reflectVal := reflect.ValueOf(args)
	reflectValType := reflectVal.Type()
	for i := 0; i < reflectValType.NumField(); i++ {
		field := reflectValType.Field(i);
		jsonFieldName := field.Tag.Get(jsonFieldTag)

		// Ensure no empty strings
		strVal := reflectVal.Field(i).String()
		if strings.TrimSpace(strVal) == "" {
			return nil, stacktrace.NewError("JSON field '%s' is whitespace or empty string", jsonFieldName)
		}
	}

	return &args, nil
}
